src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -1040,6 +1050,186 @@ prep_local_assembly (RpmOstreeSysrootUpgrader *self,
+}
+
+typedef struct {
+  guint counter;
+  GMainLoop *loop;
+  GError **error;
+} InitRamfsOverlayData;
+
+static void
+splice_cb (GObject      *object,
+           GAsyncResult *result,
+           gpointer      user_data)
+{
+  InitRamfsOverlayData *data = user_data;
+  g_autoptr(GError) local_error = NULL;
+
+  (void)g_output_stream_splice_finish ((GOutputStream*)object, result, &local_error);
+
+  /* just keep the first error */
+  if (local_error && !*data->error)
+    *data->error = g_steal_pointer (&local_error);
+
+  data->counter--;
+  if (data->counter == 0)
+    g_main_loop_quit (data->loop);
+}
+
+static int
+compare_strings (gconstpointer a,
+                 gconstpointer b)
+{
+  const gchar *sa = *(const gchar **) a;
+  const gchar *sb = *(const gchar **) b;
+  return strcmp (sa, sb);
+}
+
+static gboolean
+generate_initramfs_overlay (GHashTable    *initramfs_etc_files,
+                            GLnxTmpfile   *out_tmpf,
+                            GCancellable  *cancellable,
+                            GError       **error)
+{
+
+  g_autoptr(GMainLoop) loop = g_main_loop_new (g_main_context_get_thread_default (), TRUE);
+  g_autoptr(GError) local_error = NULL;
+  InitRamfsOverlayData data = { .loop = loop, .error = &local_error };
+
+  g_auto(GLnxTmpfile) tmpf = { 0, };
+  if (!glnx_open_anonymous_tmpfile (O_RDWR | O_CLOEXEC, &tmpf, error))
+    return glnx_prefix_error (error, "Creating tmpfile");
+
+  g_autoptr(GOutputStream) tmpf_stream = g_unix_output_stream_new (tmpf.fd, FALSE);
+
+  g_autoptr(GPtrArray) filelist = g_ptr_array_new_with_free_func (g_free);
+
+  /* could do this async too... though it really shouldn't be that many files */
+  GLNX_HASH_TABLE_FOREACH (initramfs_etc_files, const char*, path)
+    {
+      /* should've been checked already */
+      g_assert (g_str_has_prefix (path, "/etc/"));
+      if (!add_etc_files_recurse (path, filelist, cancellable, error))
+        return FALSE;
+    }
+
+  /* feed into the input stream, but sort first to avoid needlessly changing the checksum */
+  g_ptr_array_sort (filelist, compare_strings);
+  g_autoptr(GInputStream) filelist_input = g_memory_input_stream_new ();
+
+  /* keep a hash of the previous path so we don't archive the same file multiple times */
+  guint last_path_hash = 0;
+  for (guint i = 0; i < filelist->len; i++)
+    {
+      g_autofree char *path = g_steal_pointer (&filelist->pdata[i]); /* steal in-place */
+      guint path_hash = g_str_hash (path);
+      if (path_hash == last_path_hash)
+        continue;
+      g_memory_input_stream_add_data (G_MEMORY_INPUT_STREAM (filelist_input),
+                                      g_steal_pointer (&path), (strlen (path))+1, g_free);
+      last_path_hash = path_hash;
+    }
+
+  const char *cpio_argv[] = {"cpio", "--create", "--format", "newc", "--quiet",
+                             "--reproducible", "--null", NULL};
+  g_autoptr(GSubprocess) cpio = g_subprocess_newv ((const char *const*)cpio_argv,
```

Comment:
https://developer.gnome.org/gio/stable/GSubprocess.html#g-subprocess-communicate also takes stdin as a buffer, so we wouldn't need to splice anything in combination with the below.